File: skills/langchain-python/scripts/init_project.py
#!/usr/bin/env python3
"""
Initialize a new LangChain Python project from templates.

Usage:
    python init_project.py PROJECT_NAME [--template TEMPLATE]
    
Templates:
    - simple: Basic agent with tools (default)
    - langgraph: Advanced LangGraph workflow
    - rag: RAG Q&A application
"""

import os
import sys
import shutil
import argparse
from pathlib import Path

TEMPLATES = {
    "simple": "simple-agent",
    "langgraph": "langgraph-agent",
    "rag": "rag-app"
}

def create_project(project_name: str, template: str):
    """Create a new LangChain project from template."""
    
    # Get template path
    script_dir = Path(__file__).parent
    template_dir = script_dir.parent / "assets" / TEMPLATES[template]
    
    if not template_dir.exists():
        print(f"Error: Template '{template}' not found at {template_dir}")
        sys.exit(1)
    
    # Create project directory
    project_dir = Path(project_name)
    
    if project_dir.exists():
        print(f"Error: Directory '{project_name}' already exists")
        sys.exit(1)
    
    print(f"Creating project '{project_name}' from '{template}' template...")
    
    # Copy template
    shutil.copytree(template_dir, project_dir)
    
    # Create .env file
    env_file = project_dir / ".env"
    with open(env_file, "w") as f:
        f.write("""# API Keys
OPENAI_API_KEY=your-openai-api-key-here
ANTHROPIC_API_KEY=your-anthropic-api-key-here

# LangSmith (optional, for tracing)
LANGCHAIN_TRACING_V2=true
LANGCHAIN_API_KEY=your-langsmith-api-key-here
LANGCHAIN_PROJECT=my-project
""")
    
    # Create requirements.txt
    requirements = project_dir / "requirements.txt"
    with open(requirements, "w") as f:
        if template == "simple":
            f.write("""langchain==0.1.0
langchain-openai==0.0.5
langchain-community==0.0.20
python-dotenv==1.0.0
""")
        elif template == "langgraph":
            f.write("""langchain==0.1.0
langchain-openai==0.0.5
langgraph==0.0.30
python-dotenv==1.0.0
""")
        elif template == "rag":
            f.write("""langchain==0.1.0
langchain-openai==0.0.5
langchain-community==0.0.20
faiss-cpu==1.7.4
python-dotenv==1.0.0
""")
    
    # Create .gitignore
    gitignore = project_dir / ".gitignore"
    with open(gitignore, "w") as f:
        f.write("""# Environment
.env
venv/
__pycache__/
*.pyc

# Data
*.db
*.sqlite
faiss_index/
chroma_db/

# IDE
.vscode/
.idea/
""")
    
    print(f"\n✓ Project '{project_name}' created successfully!")
    print(f"\nNext steps:")
    print(f"  1. cd {project_name}")
    print(f"  2. python -m venv venv")
    print(f"  3. source venv/bin/activate  # or 'venv\\Scripts\\activate' on Windows")
    print(f"  4. pip install -r requirements.txt")
    print(f"  5. Edit .env with your API keys")
    print(f"  6. Run the main script")
    print(f"\nRead README.md for more information.")

def main():
    """Main function."""
    parser = argparse.ArgumentParser(
        description="Initialize a new LangChain Python project"
    )
    
    parser.add_argument(
        "project_name",
        help="Name of the project to create"
    )
    
    parser.add_argument(
        "--template",
        "-t",
        choices=list(TEMPLATES.keys()),
        default="simple",
        help="Template to use (default: simple)"
    )
    
    args = parser.parse_args()
    
    create_project(args.project_name, args.template)

if __name__ == "__main__":
    main()

File: skills/langchain-typescript/scripts/init_project.ts
#!/usr/bin/env node

/**
 * Initialize a new LangChain TypeScript project
 * 
 * Usage: tsx init_project.ts <project-name> [--template simple|langgraph|rag]
 */

import * as fs from "fs/promises";
import * as path from "path";

interface ProjectConfig {
  name: string;
  template: "simple" | "langgraph" | "rag";
}

const templates = {
  simple: {
    description: "Simple agent with tool calling",
    files: ["agent.ts", "package.json", "README.md"]
  },
  langgraph: {
    description: "Advanced LangGraph agent with state management",
    files: ["graph-agent.ts", "package.json", "README.md"]
  },
  rag: {
    description: "RAG application for question answering",
    files: ["rag.ts", "package.json", "README.md"]
  }
};

async function parseArgs(): Promise<ProjectConfig> {
  const args = process.argv.slice(2);
  
  if (args.length === 0 || args[0] === "--help" || args[0] === "-h") {
    console.log(`
LangChain TypeScript Project Initializer

Usage:
  tsx init_project.ts <project-name> [--template <type>]

Templates:
  simple     - Simple agent with tool calling (default)
  langgraph  - Advanced agent with LangGraph
  rag        - RAG application

Examples:
  tsx init_project.ts my-agent
  tsx init_project.ts my-app --template langgraph
  tsx init_project.ts qa-bot --template rag
    `);
    process.exit(0);
  }
  
  const name = args[0];
  const templateIndex = args.indexOf("--template");
  const template = templateIndex !== -1 && args[templateIndex + 1]
    ? args[templateIndex + 1] as "simple" | "langgraph" | "rag"
    : "simple";
  
  if (!["simple", "langgraph", "rag"].includes(template)) {
    console.error(`Invalid template: ${template}`);
    console.error("Available templates: simple, langgraph, rag");
    process.exit(1);
  }
  
  return { name, template };
}

async function copyTemplate(
  templatePath: string,
  targetPath: string
): Promise<void> {
  const files = await fs.readdir(templatePath);
  
  for (const file of files) {
    const sourcePath = path.join(templatePath, file);
    const destPath = path.join(targetPath, file);
    
    const stat = await fs.stat(sourcePath);
    
    if (stat.isDirectory()) {
      await fs.mkdir(destPath, { recursive: true });
      await copyTemplate(sourcePath, destPath);
    } else {
      await fs.copyFile(sourcePath, destPath);
    }
  }
}

async function createProject(config: ProjectConfig): Promise<void> {
  const { name, template } = config;
  
  console.log(`\n🚀 Creating LangChain TypeScript project: ${name}`);
  console.log(`   Template: ${template} - ${templates[template].description}\n`);
  
  // Create project directory
  const projectPath = path.join(process.cwd(), name);
  
  try {
    await fs.access(projectPath);
    console.error(`❌ Error: Directory '${name}' already exists`);
    process.exit(1);
  } catch {
    // Directory doesn't exist, proceed
  }
  
  await fs.mkdir(projectPath, { recursive: true });
  console.log(`✅ Created directory: ${name}`);
  
  // Determine template source path
  const scriptDir = path.dirname(new URL(import.meta.url).pathname);
  const templatePath = path.join(scriptDir, "..", "assets", `${template}-agent`);
  
  // Copy template files
  await copyTemplate(templatePath, projectPath);
  console.log(`✅ Copied template files`);
  
  // Create additional config files
  await createConfigFiles(projectPath);
  console.log(`✅ Created configuration files`);
  
  // Create .gitignore
  await fs.writeFile(
    path.join(projectPath, ".gitignore"),
    `node_modules/
dist/
.env
.env.local
*.log
.DS_Store
`
  );
  console.log(`✅ Created .gitignore`);
  
  // Create .env.example
  await fs.writeFile(
    path.join(projectPath, ".env.example"),
    `OPENAI_API_KEY=your_openai_api_key_here
ANTHROPIC_API_KEY=your_anthropic_api_key_here
LANGSMITH_API_KEY=your_langsmith_api_key_here
LANGSMITH_TRACING_V2=true
LANGSMITH_PROJECT=${name}
`
  );
  console.log(`✅ Created .env.example`);
  
  console.log(`\n✨ Project created successfully!\n`);
  console.log(`Next steps:`);
  console.log(`  1. cd ${name}`);
  console.log(`  2. npm install`);
  console.log(`  3. cp .env.example .env`);
  console.log(`  4. Edit .env and add your API keys`);
  console.log(`  5. npm run dev\n`);
}

async function createConfigFiles(projectPath: string): Promise<void> {
  // tsconfig.json
  const tsconfig = {
    compilerOptions: {
      target: "ES2022",
      module: "ESNext",
      moduleResolution: "bundler",
      esModuleInterop: true,
      strict: true,
      skipLibCheck: true,
      resolveJsonModule: true,
      outDir: "./dist",
      rootDir: "./",
      declaration: true,
      declarationMap: true,
      sourceMap: true
    },
    include: ["**/*.ts"],
    exclude: ["node_modules", "dist"]
  };
  
  await fs.writeFile(
    path.join(projectPath, "tsconfig.json"),
    JSON.stringify(tsconfig, null, 2)
  );
  
  // .editorconfig
  const editorconfig = `root = true

[*]
indent_style = space
indent_size = 2
end_of_line = lf
charset = utf-8
trim_trailing_whitespace = true
insert_final_newline = true

[*.md]
trim_trailing_whitespace = false
`;
  
  await fs.writeFile(
    path.join(projectPath, ".editorconfig"),
    editorconfig
  );
}

// Main execution
async function main() {
  try {
    const config = await parseArgs();
    await createProject(config);
  } catch (error) {
    console.error(`\n❌ Error: ${(error as Error).message}`);
    process.exit(1);
  }
}

main();
